pkg/extensions/sync/sync.go

```diff
@@ -319,186 +326,227 @@ func syncRegistry(ctx context.Context, regCfg RegistryConfig, upstreamURL string
 
 	log.Info().Msgf("got repos: %v", repos)
 
-	var images []struct {
+	upstreamAddr := StripRegistryTransport(upstreamURL)
+
+	reposWithContentID := make(map[string][]struct {
 		ref     types.ImageReference
 		content Content
-	}
-
-	upstreamAddr := StripRegistryTransport(upstreamURL)
+	})
 
 	for contentID, repos := range repos {
 		r := repos
 		contentID := contentID
 
 		if err = retry.RetryIfNecessary(ctx, func() error {
-			refs, err := imagesToCopyFromUpstream(ctx, upstreamAddr, r, upstreamCtx, regCfg.Content[contentID], log)
-			for _, ref := range refs {
-				images = append(images, struct {
-					ref     types.ImageReference
-					content Content
-				}{
-					ref:     ref,
-					content: regCfg.Content[contentID],
-				})
+			for _, repo := range r {
+				refs, err := imagesToCopyFromUpstream(ctx, upstreamAddr, r, upstreamCtx, regCfg.Content[contentID], log)
+				if err != nil {
+					return err
+				}
+
+				for _, ref := range refs[repo] {
+					reposWithContentID[repo] = append(reposWithContentID[repo], struct {
+						ref     types.ImageReference
+						content Content
+					}{
+						ref:     ref,
+						content: regCfg.Content[contentID],
+					})
+				}
 			}
 
-			return err
+			return nil
 		}, retryOptions); err != nil {
 			log.Error().Err(err).Msg("error while getting images references from upstream, retrying...")
 
 			return err
 		}
 	}
 
-	for _, image := range images {
+	for _, imageList := range reposWithContentID {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		default:
 			break
 		}
 
-		upstreamImageRef := image.ref
+		remoteRepo := getRepoFromRef(imageList[0].ref, upstreamAddr)
```

Comment:
no need for this, you already have it in reposWithContentID

just write it like: for remoteRepo, imageList := range reposWithContentID  {}